### PR TITLE
feat: preserve related_name on Observations

### PR DESCRIPTION
**Pull request by miketheman**
If we want to allow ourselves to preserve some details about a given
Observation after the related model has been removed, copy some of the
data on to the Observation record.

Since we can infer the model type from the table name, we don't need to
store that.

Contains a data migration to update existing records and enforce the
requirement at the database level.

Signed-off-by: Mike Fiedler <miketheman@gmail.com>

### CODE REVIEW

File: warehouse/packaging/models.py [hunk 1]
```diff
@@ -622,6 +622,16 @@ def __table_args__(cls):  # noqa
...
 
+    @property
+    def name(self) -> str:
+        """
+        Release doesn't have a name column, inherit from its Project.
+
+        NOTE: `name` is used in `warehouse.packaging.tasks.sync_bigquery_release_files`
+        If changed here and not there, may have implications on downstream data.
+        """
+        return self.project.name
```
Comment (ewdurbin) [inline warehouse/packaging/models.py:633]: should this (or the property suggested above) perhaps be `f”{self.project.name}@{self.version}`? or similar to not discard too much fidelity?

File: warehouse/observations/models.py [hunk 2]
```diff
@@ -244,6 +249,7 @@ def record_observation(
             observer=actor.observer,
             payload=payload,
             related=self,
```
Comment (ewdurbin) [inline warehouse/observations/models.py:251]: feels bad to be implicitly binding to `name`, should we perhaps use an explicit `related_name` property for models we may denormalize as such?

Comment (miketheman) [reply warehouse/observations/models.py:251]: I tried to use the pre-exisitng `__repr__()` call that these models implement, but it didn't exactly look right.

The resulting strings look like: `"Project(name='example')"` for a Project, and `"Release(project=Project(name='example'), version='0.0.2')"`.

These are semi-seralized, as in we could feasibly turn them back into structured data, but they likely won't hold up to efficient SQL queries - but maybe they don't have to?

Comment (miketheman) [reply warehouse/observations/models.py:251]: I've updated the scope to use `repr()` for these, and dropped the bind to `name`.

File: warehouse/packaging/models.py (hunk 1 above)
Comment (miketheman) [reply warehouse/packaging/models.py]: Dropped this scope, and using `repr()` instead.

**Review by ewdurbin — APPROVED** [review]
probably fine as is, but i think the raised comment regarding query efficiency if we ever _did_ want to do something like 'show all past observations related to the project name "widget"' would require a refactor.

what's included via repr should be enough for a refactor down the line so that's probably a "when we cross the bridge" concern.

i can't come up with a simple solution. repr is a fine compromise for now but something explicit for the purpose such as a dedicated explicit `related_name` property seems less likely to accidentally drift (not that we change our reprs.... uh ever?).

**Pull request merged**